x/stake/pool_test.go

```diff
@@ -257,21 +257,19 @@ func TestCandidateRemoveShares(t *testing.T) {
 // TODO Make all random tests less obfuscated!
 
 // generate a random candidate
-func randomCandidate(r *rand.Rand) Candidate {
+func randomCandidate(r *rand.Rand, i int) Candidate {
 	var status CandidateStatus
 	if r.Float64() < float64(0.5) {
 		status = Bonded
 	} else {
 		status = Unbonded
 	}
-	assets := sdk.NewRat(int64(r.Int31n(10000)))
-	liabilities := sdk.NewRat(int64(r.Int31n(10000)))
 	return Candidate{
 		Status:      status,
-		Address:     addrs[0],
-		PubKey:      pks[0],
-		Assets:      assets,
-		Liabilities: liabilities,
+		Address:     addrs[i], //max 40 right now, as addrs only goes up to addrs[39]
+		PubKey:      pks[i],
+		Assets:      sdk.NewRat(0),
```

Comment:
The previous way the assets and liabilities were randomized, liabilities could be higher than assets, which should't be allowed since assets = liabilities + self stake (if i am not mistaken). Now tokens are randomized, and passed into pool.candidateAddtokens() to get assets and liabilities

Reply:
You are mistaken, assets and liabilities are independant. The name has been changed in the refactor to make it a bit more clear now PoolShares and DelegationShares. A validator can give away more delegation shares than they have of pool shares it just means the exchanges rate will be less than one